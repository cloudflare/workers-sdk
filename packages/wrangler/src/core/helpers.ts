import { CommandLineArgsError } from "../errors";
import type { CreateCommandResult } from "./create-command";
import type {
	AliasDefinition,
	CommandDefinition,
	NamedArgDefinitions,
	NamespaceDefinition,
} from "./types";

/**
 * A helper to demand one of a set of options
 * via https://github.com/yargs/yargs/issues/1093#issuecomment-491299261
 */
export function demandOneOfOption(...options: string[]) {
	return function (argv: { [key: string]: unknown }) {
		const count = options.filter((option) => argv[option]).length;
		const lastOption = options.pop();

		if (count === 0) {
			throw new CommandLineArgsError(
				`Exactly one of the arguments ${options.join(
					", "
				)} and ${lastOption} is required`
			);
		} else if (count > 1) {
			throw new CommandLineArgsError(
				`Arguments ${options.join(
					", "
				)} and ${lastOption} are mutually exclusive`
			);
		}

		return true;
	};
}

/**
 * A helper to ensure that an argument only receives a single value.
 *
 * This is a workaround for a limitation in yargs where non-array arguments can still receive multiple values
 * even though the inferred type is not an array.
 *
 * @see https://github.com/yargs/yargs/issues/1318
 */
export function demandSingleValue<Argv extends { [key: string]: unknown }>(
	key: string,
	allow?: (argv: Argv) => boolean
) {
	return function (argv: Argv) {
		if (Array.isArray(argv[key]) && !allow?.(argv)) {
			throw new CommandLineArgsError(
				`The argument "--${key}" expects a single value, but received multiple: ${JSON.stringify(argv[key])}.`
			);
		}

		return true;
	};
}

/**
 * Checks if a definition is an alias definition.
 */
export function isAliasDefinition(
	def:
		| AliasDefinition
		| CreateCommandResult<NamedArgDefinitions>
		| NamespaceDefinition
): def is AliasDefinition {
	return (def as AliasDefinition).aliasOf !== undefined;
}

/**
 * Checks if a definition is a command definition.
 */
export function isCommandDefinition(
	def:
		| AliasDefinition
		| CreateCommandResult<NamedArgDefinitions>
		| NamespaceDefinition
): def is CommandDefinition {
	return (def as CommandDefinition).handler !== undefined;
}

/**
 * Checks if a definition is a namespace definition.
 */
export function isNamespaceDefinition(
	def:
		| AliasDefinition
		| CreateCommandResult<NamedArgDefinitions>
		| NamespaceDefinition
): def is NamespaceDefinition {
	return !isAliasDefinition(def) && !isCommandDefinition(def);
}
