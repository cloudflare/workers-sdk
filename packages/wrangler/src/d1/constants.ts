export const DEFAULT_MIGRATION_PATH = "./migrations";
export const DEFAULT_MIGRATION_TABLE = "d1_migrations";
