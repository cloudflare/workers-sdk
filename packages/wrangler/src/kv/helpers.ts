import { URLSearchParams } from "node:url";
import { FormData } from "undici";
import { fetchListResult, fetchResult, fetchKVGetValue } from "../cfetch";
import { logger } from "../logger";
import type { Config } from "../config";

/** The largest number of kv items we can pass to the API in a single request. */
const API_MAX = 10000;
// The const below are halved from the API's true capacity to help avoid
// hammering it with large requests.
export const BATCH_KEY_MAX = API_MAX / 2;

type KvArgs = {
	binding?: string;
	"namespace-id"?: string;
	preview?: boolean;
};

/**
 * Create a new namespace under the given `accountId` with the given `title`.
 *
 * @returns the generated id of the created namespace.
 */
export async function createKVNamespace(
	accountId: string,
	title: string
): Promise<string> {
	const response = await fetchResult<{ id: string }>(
		`/accounts/${accountId}/storage/kv/namespaces`,
		{
			method: "POST",
			headers: {
				"Content-Type": "application/json",
			},
			body: JSON.stringify({
				title,
			}),
		}
	);

	return response.id;
}

/**
 * The information about a namespace that is returned from `listNamespaces()`.
 */
export interface KVNamespaceInfo {
	id: string;
	title: string;
	supports_url_encoding?: boolean;
}

/**
 * Fetch a list of all the namespaces under the given `accountId`.
 */
export async function listKVNamespaces(
	accountId: string
): Promise<KVNamespaceInfo[]> {
	const pageSize = 100;
	let page = 1;
	const results: KVNamespaceInfo[] = [];
	while (results.length % pageSize === 0) {
		const json = await fetchResult<KVNamespaceInfo[]>(
			`/accounts/${accountId}/storage/kv/namespaces`,
			{},
			new URLSearchParams({
				per_page: pageSize.toString(),
				order: "title",
				direction: "asc",
				page: page.toString(),
			})
		);
		page++;
		results.push(...json);
		if (json.length < pageSize) {
			break;
		}
	}
	return results;
}

export interface NamespaceKeyInfo {
	name: string;
	expiration?: number;
	metadata?: { [key: string]: unknown };
}

export async function listKVNamespaceKeys(
	accountId: string,
	namespaceId: string,
	prefix = ""
) {
	return await fetchListResult<NamespaceKeyInfo>(
		`/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/keys`,
		{},
		new URLSearchParams({ prefix })
	);
}

export async function deleteKVNamespace(
	accountId: string,
	namespaceId: string
) {
	return await fetchResult<{ id: string }>(
		`/accounts/${accountId}/storage/kv/namespaces/${namespaceId}`,
		{ method: "DELETE" }
	);
}

/**
 * Information about a key-value pair, including its "metadata" fields.
 */
export interface KeyValue {
	key: string;
	value: string | Buffer;
	expiration?: number;
	expiration_ttl?: number;
	metadata?: object;
	base64?: boolean;
}

const KeyValueKeys = new Set([
	"key",
	"value",
	"expiration",
	"expiration_ttl",
	"metadata",
	"base64",
]);

/**
 * The object has the specified property.
 */
function hasProperty<T extends object>(
	obj: object,
	property: keyof T
): obj is T {
	return property in obj;
}

/**
 * The object has a required property of the specified type.
 */
function hasTypedProperty<T extends object>(
	obj: object,
	property: keyof T,
	type: string
): obj is T {
	return hasProperty(obj, property) && typeof obj[property] === type;
}

/**
 * The object an optional property, of the specified type.
 */
function hasOptionalTypedProperty<T extends object>(
	obj: object,
	property: keyof T,
	type: string
): obj is Omit<T, typeof property> | T {
	return !hasProperty(obj, property) || typeof obj[property] === type;
}

/**
 * Is the given object a valid `KeyValue` type?
 */
export function isKVKeyValue(keyValue: unknown): keyValue is KeyValue {
	if (
		keyValue === null ||
		typeof keyValue !== "object" ||
		!hasTypedProperty(keyValue, "key", "string") ||
		!hasTypedProperty(keyValue, "value", "string") ||
		!hasOptionalTypedProperty(keyValue, "expiration", "number") ||
		!hasOptionalTypedProperty(keyValue, "expiration_ttl", "number") ||
		!hasOptionalTypedProperty(keyValue, "base64", "boolean") ||
		!hasOptionalTypedProperty(keyValue, "metadata", "object")
	) {
		return false;
	}
	return true;
}

/**
 * Get all the properties on the `keyValue` that are not expected.
 */
export function unexpectedKVKeyValueProps(keyValue: KeyValue): string[] {
	const props = Object.keys(keyValue);
	return props.filter((prop) => !KeyValueKeys.has(prop));
}

/**
 * Turn object with fields into FormData
 */
function asFormData(fields: Record<string, unknown>): FormData {
	const formData = new FormData();

	for (const [name, value] of Object.entries(fields)) {
		formData.append(name, value);
	}

	return formData;
}

export async function putKVKeyValue(
	accountId: string,
	namespaceId: string,
	keyValue: KeyValue
) {
	let searchParams: URLSearchParams | undefined;
	if (keyValue.expiration || keyValue.expiration_ttl) {
		searchParams = new URLSearchParams();
		if (keyValue.expiration) {
			searchParams.set("expiration", `${keyValue.expiration}`);
		}
		if (keyValue.expiration_ttl) {
			searchParams.set("expiration_ttl", `${keyValue.expiration_ttl}`);
		}
	}
	return await fetchResult(
		`/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/values/${encodeURIComponent(
			keyValue.key
		)}`,
		{
			method: "PUT",
			body: keyValue.metadata
				? asFormData({
						value: keyValue.value,
						metadata: JSON.stringify(keyValue.metadata),
				  })
				: keyValue.value,
		},
		searchParams
	);
}

export async function getKVKeyValue(
	accountId: string,
	namespaceId: string,
	key: string
): Promise<ArrayBuffer> {
	return await fetchKVGetValue(accountId, namespaceId, encodeURIComponent(key));
}

export async function deleteKVKeyValue(
	accountId: string,
	namespaceId: string,
	key: string
) {
	return await fetchResult(
		`/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/values/${encodeURIComponent(
			key
		)}`,
		{ method: "DELETE" }
	);
}

/**
 * Formatter for converting e.g. 5328 --> 5,328
 */
export const formatNumber = new Intl.NumberFormat("en-US", {
	notation: "standard",
}).format;

/**
 * Helper function for bulk requests, logs ongoing output to console.
 */
function logBulkProgress(
	operation: "put" | "delete",
	index: number,
	total: number
) {
	logger.log(
		`${operation === "put" ? "Uploaded" : "Deleted"} ${Math.floor(
			(100 * index) / total
		)}% (${formatNumber(index)} out of ${formatNumber(total)})`
	);
}

export async function putKVBulkKeyValue(
	accountId: string,
	namespaceId: string,
	keyValues: KeyValue[],
	quiet = false,
	abortSignal?: AbortSignal
) {
	for (let index = 0; index < keyValues.length; index += BATCH_KEY_MAX) {
		if (!quiet && keyValues.length > BATCH_KEY_MAX) {
			logBulkProgress("put", index, keyValues.length);
		}

		await fetchResult(
			`/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/bulk`,
			{
				method: "PUT",
				body: JSON.stringify(keyValues.slice(index, index + BATCH_KEY_MAX)),
				headers: { "Content-Type": "application/json" },
			},
			undefined,
			abortSignal
		);
	}

	if (!quiet && keyValues.length > BATCH_KEY_MAX) {
		logBulkProgress("put", keyValues.length, keyValues.length);
	}
}

export async function deleteKVBulkKeyValue(
	accountId: string,
	namespaceId: string,
	keys: string[],
	quiet = false
) {
	for (let index = 0; index < keys.length; index += BATCH_KEY_MAX) {
		if (!quiet && keys.length > BATCH_KEY_MAX) {
			logBulkProgress("delete", index, keys.length);
		}

		await fetchResult(
			`/accounts/${accountId}/storage/kv/namespaces/${namespaceId}/bulk`,
			{
				method: "DELETE",
				body: JSON.stringify(keys.slice(index, index + BATCH_KEY_MAX)),
				headers: { "Content-Type": "application/json" },
			}
		);
	}
	if (!quiet && keys.length > BATCH_KEY_MAX) {
		logBulkProgress("delete", keys.length, keys.length);
	}
}

export function getKVNamespaceId(
	{ preview, binding, "namespace-id": namespaceId }: KvArgs,
	config: Config
): string {
	// nice
	if (namespaceId) {
		return namespaceId;
	}

	// begin pre-flight checks

	// `--binding` is only valid if there's a wrangler configuration.
	if (binding && !config) {
		throw new Error("--binding specified, but no config file was found.");
	}

	// there's no config. abort here
	if (!config) {
		throw new Error(
			"Failed to find a config file.\n" +
				"Either use --namespace-id to upload directly or create a configuration file with a binding."
		);
	}

	// there's no KV namespaces
	if (!config.kv_namespaces || config.kv_namespaces.length === 0) {
		throw new Error(
			"No KV Namespaces configured! Either use --namespace-id to upload directly or add a KV namespace to your wrangler config file."
		);
	}

	const namespace = config.kv_namespaces.find((ns) => ns.binding === binding);

	// we couldn't find a namespace with that binding
	if (!namespace) {
		throw new Error(
			`A namespace with binding name "${binding}" was not found in the configured "kv_namespaces".`
		);
	}

	// end pre-flight checks

	// we're in preview mode, `--preview true` or `--preview` was passed
	if (preview && namespace.preview_id) {
		namespaceId = namespace.preview_id;
		// We don't want to execute code below if preview is set to true, so we just return. Otherwise we will get errors!
		return namespaceId;
	} else if (preview) {
		throw new Error(
			`No preview ID found for ${binding}. Add one to your wrangler config file to use a separate namespace for previewing your worker.`
		);
	}

	// either `--preview false`, or preview wasn't passed
	// TODO: should we care? or should we just treat false and undefined the same
	const previewIsDefined = typeof preview !== "undefined";

	// --preview false was passed
	if (previewIsDefined && namespace.id) {
		namespaceId = namespace.id;
		// We don't want to execute code below if preview is set to true, so we just return. Otherwise we can get error!
		return namespaceId;
	} else if (previewIsDefined) {
		throw new Error(
			`No namespace ID found for ${binding}. Add one to your wrangler config file to use a separate namespace for previewing your worker.`
		);
	}

	// `--preview` wasn't passed
	const bindingHasOnlyOneId =
		(namespace.id && !namespace.preview_id) ||
		(!namespace.id && namespace.preview_id);
	if (bindingHasOnlyOneId) {
		namespaceId = namespace.id || namespace.preview_id;
	} else {
		throw new Error(
			`${binding} has both a namespace ID and a preview ID. Specify "--preview" or "--preview false" to avoid writing data to the wrong namespace.`
		);
	}

	// shouldn't happen. we should be able to prove this with strong typing.
	// TODO: when we add strongly typed commands, rewrite these checks so they're exhaustive
	if (!namespaceId) {
		throw Error(
			"Something went wrong trying to determine which namespace to upload to.\n" +
				"Please create a github issue with the command you just ran along with your wrangler configuration."
		);
	}

	return namespaceId;
}

/**
 * KV namespace binding names must be valid JS identifiers.
 */
export function isValidKVNamespaceBinding(
	binding: string | undefined
): binding is string {
	return (
		typeof binding === "string" && /^[a-zA-Z_][a-zA-Z0-9_]*$/.test(binding)
	);
}
