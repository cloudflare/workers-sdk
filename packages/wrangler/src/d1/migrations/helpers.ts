import fs from "node:fs";
import path from "path";
import { confirm } from "../../dialogs";
import { CI } from "../../is-ci";
import isInteractive from "../../is-interactive";
import { logger } from "../../logger";
import { DEFAULT_MIGRATION_PATH } from "../constants";
import { executeSql } from "../execute";
import type { ConfigFields, DevConfig, Environment } from "../../config";
import type { QueryResult } from "../execute";
import type { Migration } from "../types";

export async function getMigrationsPath({
	projectPath,
	migrationsFolderPath,
	createIfMissing,
}: {
	projectPath: string;
	migrationsFolderPath: string;
	createIfMissing: boolean;
}): Promise<string> {
	const dir = path.resolve(projectPath, migrationsFolderPath);
	if (fs.existsSync(dir)) return dir;

	const warning = `No migrations folder found.${
		migrationsFolderPath === DEFAULT_MIGRATION_PATH
			? " Set `migrations_dir` in wrangler.toml to choose a different path."
			: ""
	}`;

	if (createIfMissing && (await confirm(`${warning}\nOk to create ${dir}?`))) {
		fs.mkdirSync(dir, { recursive: true });
		return dir;
	} else {
		logger.warn(warning);
	}

	throw new Error(`No migrations present at ${dir}.`);
}

export async function getUnappliedMigrations({
	migrationsTableName,
	migrationsPath,
	local,
	config,
	name,
	persistTo,
	preview,
}: {
	migrationsTableName: string;
	migrationsPath: string;
	local: boolean | undefined;
	config: ConfigFields<DevConfig> & Environment;
	name: string;
	persistTo: string | undefined;
	preview: boolean | undefined;
}): Promise<Array<string>> {
	const appliedMigrations = (
		await listAppliedMigrations(
			migrationsTableName,
			local,
			config,
			name,
			persistTo,
			preview
		)
	).map((migration) => {
		return migration.name;
	});
	const projectMigrations = getMigrationNames(migrationsPath);

	const unappliedMigrations: Array<string> = [];

	for (const migration of projectMigrations) {
		if (!appliedMigrations.includes(migration)) {
			unappliedMigrations.push(migration);
		}
	}

	return unappliedMigrations;
}

const listAppliedMigrations = async (
	migrationsTableName: string,
	local: boolean | undefined,
	config: ConfigFields<DevConfig> & Environment,
	name: string,
	persistTo: string | undefined,
	preview: boolean | undefined
): Promise<Migration[]> => {
	const response: QueryResult[] | null = await executeSql({
		local,
		config,
		name,
		shouldPrompt: isInteractive() && !CI.isCI(),
		persistTo,
		command: `SELECT *
		FROM ${migrationsTableName}
		ORDER BY id`,
		file: undefined,
		json: undefined,
		preview,
	});

	if (!response || response[0].results.length === 0) return [];

	return response[0].results as Migration[];
};

function getMigrationNames(migrationsPath: string): Array<string> {
	const migrations = [];

	const dir = fs.opendirSync(migrationsPath);

	let dirent;
	while ((dirent = dir.readSync()) !== null) {
		if (dirent.name.endsWith(".sql")) migrations.push(dirent.name);
	}

	dir.closeSync();

	return migrations;
}

export function getNextMigrationNumber(migrationsPath: string): number {
	let highestMigrationNumber = -1;

	for (const migration in getMigrationNames(migrationsPath)) {
		const migrationNumber = parseInt(migration.split("_")[0]);

		if (migrationNumber > highestMigrationNumber) {
			highestMigrationNumber = migrationNumber;
		}
	}

	return highestMigrationNumber + 1;
}

export const initMigrationsTable = async ({
	migrationsTableName,
	local,
	config,
	name,
	persistTo,
	preview,
}: {
	migrationsTableName: string;
	local: boolean | undefined;
	config: ConfigFields<DevConfig> & Environment;
	name: string;
	persistTo: string | undefined;
	preview: boolean | undefined;
}) => {
	return executeSql({
		local,
		config,
		name,
		shouldPrompt: isInteractive() && !CI.isCI(),
		persistTo,
		command: `CREATE TABLE IF NOT EXISTS ${migrationsTableName}(
		id         INTEGER PRIMARY KEY AUTOINCREMENT,
		name       TEXT UNIQUE,
		applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
);`,
		file: undefined,
		json: undefined,
		preview,
	});
};
