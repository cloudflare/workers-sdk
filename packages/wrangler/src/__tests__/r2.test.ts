import * as fs from "node:fs";
import { rest } from "msw";
import prettyBytes from "pretty-bytes";
import { MAX_UPLOAD_SIZE } from "../r2/constants";
import { mockAccountId, mockApiToken } from "./helpers/mock-account-id";
import { mockConsoleMethods } from "./helpers/mock-console";
import { createFetchResult, msw, mswSuccessR2handlers } from "./helpers/msw";
import { runInTempDir } from "./helpers/run-in-tmp";
import { runWrangler } from "./helpers/run-wrangler";
import type { R2BucketInfo } from "../r2/helpers";

describe("r2", () => {
	const std = mockConsoleMethods();
	beforeEach(() => msw.use(...mswSuccessR2handlers));

	mockAccountId();
	mockApiToken();
	runInTempDir();

	describe("bucket", () => {
		it("should show the correct help when an invalid command is passed", async () => {
			await expect(() =>
				runWrangler("r2 bucket foo")
			).rejects.toThrowErrorMatchingInlineSnapshot(`"Unknown argument: foo"`);
			expect(std.err).toMatchInlineSnapshot(`
			          "[31mX [41;31m[[41;97mERROR[41;31m][0m [1mUnknown argument: foo[0m

			          "
		        `);
			expect(std.out).toMatchInlineSnapshot(`
			"
			wrangler r2 bucket

			Manage R2 buckets

			Commands:
			  wrangler r2 bucket create <name>  Create a new R2 bucket
			  wrangler r2 bucket list           List R2 buckets
			  wrangler r2 bucket delete <name>  Delete an R2 bucket

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
		});

		describe("list", () => {
			it("should list buckets & check request inputs", async () => {
				const expectedBuckets: R2BucketInfo[] = [
					{ name: "bucket-1-local-once", creation_date: "01-01-2001" },
					{ name: "bucket-2-local-once", creation_date: "01-01-2001" },
				];
				msw.use(
					rest.get(
						"*/accounts/:accountId/r2/buckets",
						async (request, response, context) => {
							const { accountId } = request.params;
							expect(accountId).toEqual("some-account-id");
							expect(await request.text()).toEqual("");
							return response.once(
								context.json(
									createFetchResult({
										buckets: [
											{
												name: "bucket-1-local-once",
												creation_date: "01-01-2001",
											},
											{
												name: "bucket-2-local-once",
												creation_date: "01-01-2001",
											},
										],
									})
								)
							);
						}
					)
				);
				await runWrangler("r2 bucket list");

				expect(std.err).toMatchInlineSnapshot(`""`);
				const buckets = JSON.parse(std.out);
				expect(buckets).toEqual(expectedBuckets);
			});
		});

		describe("create", () => {
			it("should error if no bucket name is given", async () => {
				await expect(
					runWrangler("r2 bucket create")
				).rejects.toThrowErrorMatchingInlineSnapshot(
					`"Not enough non-option arguments: got 0, need at least 1"`
				);
				expect(std.out).toMatchInlineSnapshot(`
			"
			wrangler r2 bucket create <name>

			Create a new R2 bucket

			Positionals:
			  name  The name of the new bucket  [string] [required]

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
				expect(std.err).toMatchInlineSnapshot(`
				            "[31mX [41;31m[[41;97mERROR[41;31m][0m [1mNot enough non-option arguments: got 0, need at least 1[0m

				            "
			          `);
			});

			it("should error if the bucket to create contains spaces", async () => {
				await expect(
					runWrangler("r2 bucket create abc def ghi")
				).rejects.toThrowErrorMatchingInlineSnapshot(
					`"Unknown arguments: def, ghi"`
				);
				expect(std.out).toMatchInlineSnapshot(`
			"
			wrangler r2 bucket create <name>

			Create a new R2 bucket

			Positionals:
			  name  The name of the new bucket  [string] [required]

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
				expect(std.err).toMatchInlineSnapshot(`
				            "[31mX [41;31m[[41;97mERROR[41;31m][0m [1mUnknown arguments: def, ghi[0m

				            "
			          `);
			});

			it("should create a bucket & check request inputs", async () => {
				msw.use(
					rest.post(
						"*/accounts/:accountId/r2/buckets",
						async (request, response, context) => {
							const { accountId } = request.params;
							expect(accountId).toEqual("some-account-id");
							expect(await request.json()).toEqual({ name: "testBucket" });
							return response.once(context.json(createFetchResult({})));
						}
					)
				);
				await runWrangler("r2 bucket create testBucket");
				expect(std.out).toMatchInlineSnapshot(`
				            "Creating bucket testBucket.
				            Created bucket testBucket."
			          `);
			});
		});

		describe("delete", () => {
			it("should error if no bucket name is given", async () => {
				await expect(
					runWrangler("r2 bucket delete")
				).rejects.toThrowErrorMatchingInlineSnapshot(
					`"Not enough non-option arguments: got 0, need at least 1"`
				);
				expect(std.out).toMatchInlineSnapshot(`
			"
			wrangler r2 bucket delete <name>

			Delete an R2 bucket

			Positionals:
			  name  The name of the bucket to delete  [string] [required]

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
				expect(std.err).toMatchInlineSnapshot(`
				            "[31mX [41;31m[[41;97mERROR[41;31m][0m [1mNot enough non-option arguments: got 0, need at least 1[0m

				            "
			          `);
			});

			it("should error if the bucket name to delete contains spaces", async () => {
				await expect(
					runWrangler("r2 bucket delete abc def ghi")
				).rejects.toThrowErrorMatchingInlineSnapshot(
					`"Unknown arguments: def, ghi"`
				);
				expect(std.out).toMatchInlineSnapshot(`
			"
			wrangler r2 bucket delete <name>

			Delete an R2 bucket

			Positionals:
			  name  The name of the bucket to delete  [string] [required]

			Flags:
			  -j, --experimental-json-config  Experimental: Support wrangler.json  [boolean]
			  -c, --config                    Path to .toml configuration file  [string]
			  -e, --env                       Environment to use for operations and .env files  [string]
			  -h, --help                      Show help  [boolean]
			  -v, --version                   Show version number  [boolean]"
		`);
				expect(std.err).toMatchInlineSnapshot(`
				            "[31mX [41;31m[[41;97mERROR[41;31m][0m [1mUnknown arguments: def, ghi[0m

				            "
			          `);
			});

			it("should delete a bucket specified by name & check requests inputs", async () => {
				msw.use(
					rest.delete(
						"*/accounts/:accountId/r2/buckets/:bucketName",
						async (request, response, context) => {
							const { accountId, bucketName } = request.params;
							expect(accountId).toEqual("some-account-id");
							expect(bucketName).toEqual("some-bucket");
							expect(await request.text()).toEqual("");
							expect(request.headers.get("authorization")).toEqual(
								"Bearer some-api-token"
							);

							return response.once(context.json(createFetchResult(null)));
						}
					)
				);
				await runWrangler(`r2 bucket delete some-bucket`);
				expect(std.out).toMatchInlineSnapshot(`
				"Deleting bucket some-bucket.
				Deleted bucket some-bucket."
			`);
			});
		});
	});

	describe("r2 object", () => {
		it("should download R2 object from bucket", async () => {
			await runWrangler(
				`r2 object get bucketName-object-test/wormhole-img.png --file ./wormhole-img.png`
			);

			expect(std.out).toMatchInlineSnapshot(`
			"Downloading \\"wormhole-img.png\\" from \\"bucketName-object-test\\".
			Download complete."
		`);
		});

		it("should upload R2 object to bucket", async () => {
			fs.writeFileSync("wormhole-img.png", "passageway");
			await runWrangler(
				`r2 object put bucketName-object-test/wormhole-img.png --file ./wormhole-img.png`
			);

			expect(std.out).toMatchInlineSnapshot(`
			"Creating object \\"wormhole-img.png\\" in bucket \\"bucketName-object-test\\".
			Upload complete."
		`);
		});

		it("should fail to upload R2 object to bucket if too large", async () => {
			const TOO_BIG_FILE_SIZE = MAX_UPLOAD_SIZE + 1024 * 1024;
			fs.writeFileSync("wormhole-img.png", Buffer.alloc(TOO_BIG_FILE_SIZE));
			await expect(
				runWrangler(
					`r2 object put bucketName-object-test/wormhole-img.png --file ./wormhole-img.png`
				)
			).rejects.toThrowErrorMatchingInlineSnapshot(`
			"Error: Wrangler only supports uploading files up to ${prettyBytes(
				MAX_UPLOAD_SIZE
			)} in size
			wormhole-img.png is ${prettyBytes(TOO_BIG_FILE_SIZE)} in size"
		`);
		});

		it("should pass all fetch option flags into requestInit & check request inputs", async () => {
			msw.use(
				rest.put(
					"*/accounts/:accountId/r2/buckets/:bucketName/objects/:objectName",
					(request, response, context) => {
						const { accountId, bucketName, objectName } = request.params;
						expect(accountId).toEqual("some-account-id");
						expect(bucketName).toEqual("bucketName-object-test");
						expect(objectName).toEqual("wormhole-img.png");
						const headersObject = request.headers.all();
						delete headersObject["user-agent"];
						expect(headersObject).toMatchInlineSnapshot(`
					Object {
					  "accept": "*/*",
					  "accept-encoding": "gzip,deflate",
					  "authorization": "Bearer some-api-token",
					  "cache-control": "cache-control-mock",
					  "connection": "close",
					  "content-disposition": "content-disposition-mock",
					  "content-encoding": "content-encoding-mock",
					  "content-language": "content-lang-mock",
					  "content-length": "10",
					  "content-type": "content-type-mock",
					  "expires": "expire-time-mock",
					  "host": "api.cloudflare.com",
					}
				`);
						return response.once(
							context.json(
								createFetchResult({
									accountId: "some-account-id",
									bucketName: "bucketName-object-test",
									objectName: "wormhole-img.png",
								})
							)
						);
					}
				)
			);
			fs.writeFileSync("wormhole-img.png", "passageway");
			const flags =
				"--ct content-type-mock --cd content-disposition-mock --ce content-encoding-mock --cl content-lang-mock --cc cache-control-mock --e expire-time-mock";

			await runWrangler(
				`r2 object put bucketName-object-test/wormhole-img.png ${flags} --file wormhole-img.png`
			);

			expect(std.out).toMatchInlineSnapshot(`
			"Creating object \\"wormhole-img.png\\" in bucket \\"bucketName-object-test\\".
			Upload complete."
		`);
		});

		it("should delete R2 object from bucket", async () => {
			await runWrangler(
				`r2 object delete bucketName-object-test/wormhole-img.png`
			);

			expect(std.out).toMatchInlineSnapshot(`
			"Deleting object \\"wormhole-img.png\\" from bucket \\"bucketName-object-test\\".
			Delete complete."
		`);
		});

		it("should not allow `--pipe` & `--file` to run together", async () => {
			fs.writeFileSync("wormhole-img.png", "passageway");
			await expect(
				runWrangler(
					`r2 object put bucketName-object-test/wormhole-img.png --pipe --file wormhole-img.png`
				)
			).rejects.toThrowErrorMatchingInlineSnapshot(
				`"Arguments pipe and file are mutually exclusive"`
			);

			expect(std.err).toMatchInlineSnapshot(`
			"[31mX [41;31m[[41;97mERROR[41;31m][0m [1mArguments pipe and file are mutually exclusive[0m

			"
		`);
		});
	});
});
