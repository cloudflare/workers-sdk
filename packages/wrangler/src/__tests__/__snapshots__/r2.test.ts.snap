// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`r2 > bucket > create > wrangler.json > should create a bucket & check request inputs 1`] = `
"Creating bucket 'testBucket'...
✅ Created bucket 'testBucket' with default storage class of Standard.

Configure your Worker to write objects to this bucket:

{
  \\"r2_buckets\\": [
    {
      \\"bucket_name\\": \\"testBucket\\",
      \\"binding\\": \\"testBucket\\"
    }
  ]
}"
`;

exports[`r2 > bucket > create > wrangler.json > should create a bucket with the expected default storage class 1`] = `
"Creating bucket 'testBucket'...
✅ Created bucket 'testBucket' with default storage class of InfrequentAccess.

Configure your Worker to write objects to this bucket:

{
  \\"r2_buckets\\": [
    {
      \\"bucket_name\\": \\"testBucket\\",
      \\"binding\\": \\"testBucket\\"
    }
  ]
}"
`;

exports[`r2 > bucket > create > wrangler.json > should create a bucket with the expected jurisdiction 1`] = `
"Creating bucket 'testBucket (eu)'...
✅ Created bucket 'testBucket (eu)' with default storage class of Standard.

Configure your Worker to write objects to this bucket:

{
  \\"r2_buckets\\": [
    {
      \\"bucket_name\\": \\"testBucket\\",
      \\"binding\\": \\"testBucket\\"
    }
  ]
}"
`;

exports[`r2 > bucket > create > wrangler.json > should create a bucket with the expected location hint 1`] = `
"Creating bucket 'testBucket'...
✅ Created bucket 'testBucket' with location hint weur and default storage class of Standard.

Configure your Worker to write objects to this bucket:

{
  \\"r2_buckets\\": [
    {
      \\"bucket_name\\": \\"testBucket\\",
      \\"binding\\": \\"testBucket\\"
    }
  ]
}"
`;

exports[`r2 > bucket > create > wrangler.toml > should create a bucket & check request inputs 1`] = `
"Creating bucket 'testBucket'...
✅ Created bucket 'testBucket' with default storage class of Standard.

Configure your Worker to write objects to this bucket:

[[r2_buckets]]
bucket_name = \\"testBucket\\"
binding = \\"testBucket\\"
"
`;

exports[`r2 > bucket > create > wrangler.toml > should create a bucket with the expected default storage class 1`] = `
"Creating bucket 'testBucket'...
✅ Created bucket 'testBucket' with default storage class of InfrequentAccess.

Configure your Worker to write objects to this bucket:

[[r2_buckets]]
bucket_name = \\"testBucket\\"
binding = \\"testBucket\\"
"
`;

exports[`r2 > bucket > create > wrangler.toml > should create a bucket with the expected jurisdiction 1`] = `
"Creating bucket 'testBucket (eu)'...
✅ Created bucket 'testBucket (eu)' with default storage class of Standard.

Configure your Worker to write objects to this bucket:

[[r2_buckets]]
bucket_name = \\"testBucket\\"
binding = \\"testBucket\\"
"
`;

exports[`r2 > bucket > create > wrangler.toml > should create a bucket with the expected location hint 1`] = `
"Creating bucket 'testBucket'...
✅ Created bucket 'testBucket' with location hint weur and default storage class of Standard.

Configure your Worker to write objects to this bucket:

[[r2_buckets]]
bucket_name = \\"testBucket\\"
binding = \\"testBucket\\"
"
`;
